Add CarbonBlack downloader

austinbyers opened the pull request:
## Overview
Size: Extra Large

CarbonBlack automatically uploads new binaries that it finds on endpoints; users who have CarbonBlack can now optionally enable a CarbonBlack downloader Lambda function to copy binaries from CarbonBlack into BinaryAlert.

The downloader can be enabled by running the new `python3 manage.py configure` command to allow the user to set the CarbonBlack URL and encrypt its API key.

Additionally, `python3 manage.py cb_copy_all` allows users to copy the entire CarbonBlack binary corpus into BinaryAlert in one go.

Full documentation will be added in the next PR.

## Change Summary
- Adds `lambda_functions/downloader`
    - Added a `enable_carbon_black_downloader` Terraform variable. All downloader resources are created only if the downloader is explicitly enabled.
- Changes to `manage.py`:
    - Adds `configure` and `cb_copy_all` commands
    - Renames `test` command to `unit_test` (to distinguish it from `live_test`)
    - Refactored configuration management into a separate class
    - Adds unit test coverage for the first time
- Pylint has been slightly relaxed to allow longer variable names and more branches
- Travis email notifications disabled
- Updates pip requirements to their latest versions
- Includes type annotations in all new code

Resolves:  #29 (add downloader)
Resolves:  #48 (additional name prefix validation)
Contributes to: #34 (type annotations)

## Tested
### CI
Added unit tests for downloader code as well as most of `manage.py`. As you can see from the commit history, mocking everything correctly was a huge pain. In the future, I think we should remove `moto` entirely (we already have to do our own Dynamo and S3 mocks due to 2 separate moto issues)

### Test Deploy: Downloader Disabled (Default)
```
$ python3 manage.py deploy
ERROR: name_prefix "" does not match format [a-z][a-z0-9_]{3,50}
Please run "python3 manage.py configure"

$ python3 manage.py configure
AWS Region (us-east-1):
Unique name prefix, e.g. "company_team": ba_test_638
Enable the CarbonBlack downloader [yes/no]? (no): no
Updated configuration successfully saved to terraform/terraform.tfvars!

$ python3 manage.py deploy
...
Apply complete! Resources: 41 added, 0 changed, 0 destroyed.

$ python3 manage.py live_test
Live test succeeded!

$ python3 manage.py cb_copy_all
ERROR: CarbonBlack downloader is not enabled.
Please run "python3 manage.py configure"
```

### Test Deploy: Enable Downloader
After the previous deploy, we can easily re-configure and re-deploy:
```
$ python3 manage.py configure
AWS Region (us-east-1):
Unique name prefix, e.g. "company_team" (ba_test_638):
Enable the CarbonBlack downloader [yes/no]? (no): yes
CarbonBlack URL: [URL redacted]
CarbonBlack API token (only needs binary read access):
Terraforming KMS key...
aws_kms_key.carbon_black_credentials: Creation complete
aws_kms_alias.encrypt_credentials_alias: Creation complete
Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
Encrypting API token...
Updated configuration successfully saved to terraform/terraform.tfvars!

$ python3 manage.py deploy
Apply complete! Resources: 8 added

$ python3 manage.py cb_copy_all
[Sampled Output]
2017-08-28 18:49:39,630 INFO   Start 32 consumers
2017-08-28 18:49:40,042 DEBUG  Enqueuing CopyTask [#0] MD5: ...
2017-08-28 18:49:40,045 DEBUG  Enqueuing CopyTask [#1] MD5: ...
2017-08-28 18:49:40,046 DEBUG  Enqueuing CopyTask [#2] MD5: ...
2017-08-28 18:49:40,046 INFO   [Consumer-1] Executing CopyTask [#0] MD5: ...
2017-08-28 18:49:40,047 INFO   [Consumer-2] Executing CopyTask [#1] MD5: ...
2017-08-28 18:49:40,047 INFO   [Consumer-3] Executing CopyTask [#2] MD5: ...
... More logs ...
2017-08-28 18:49:55,089 INFO   [Consumer-1] Exiting
2017-08-28 18:49:57,675 INFO   [Consumer-2] Exiting
2017-08-28 18:50:07,184 INFO   [Consumer-3] Exiting
2017-08-28 18:50:07,184 INFO   All CopyTasks Finished!
```

## Reviewers
to: @ryandeivert 
cc: @mime-frame @airbnb/binaryalert-maintainers

ryandeivert commented on lambda_functions/downloader/copy_all.py:
```diff
@@ -0,0 +1,145 @@
+        self.failed_queue = failed_queue
+
+        # Each process needs its own logger to avoid race conditions.
+        self.logger = logging.getLogger(self.name)
```
As an aside, the logging package actually handles race conditions pretty well through the use of locks, etc (as seen [here](https://github.com/python/cpython/blob/master/Lib/logging/__init__.py#L201). That said, this is dependent on the `threading` package being available, and I don't think you'll have that in lambda(?). So, in normal use cases, having a separate logger for each is probably unnecessary, but is probably the safer/better solution in this case.

austinbyers replied:
This is actually a one-off script that is `not` designed to run in Lambda. It's probably confusing that it's part of the `lambda_functions` hierarchy, but I'm not sure where would be a better place to put it

austinbyers replied:
You're right; [Python logging is thread-safe](https://docs.python.org/3/library/logging.html#thread-safety) I'll remove this

ryandeivert commented on tests/manage_test.py:
```diff
@@ -254,3 +245,99 @@ def test_encrypt_cb_api_token(
+    @mock.patch.object(manage.Manager, 'build')
+    @mock.patch.object(manage.Manager, 'apply')
+    @mock.patch.object(manage.Manager, 'analyze_all')
+    def test_deploy(self, mock_analyze: mock.MagicMock, mock_apply: mock.MagicMock,
```
stacks on stacks on stacks

austinbyers replied:
boots and cats on stacks on stacks

austinbyers commented on lambda_functions/downloader/main.py:
```diff
@@ -0,0 +1,125 @@
+    """Upload a binary to S3, keyed by a UUID.
+
+    Args:
+        local_file_path: [string] Path to the file to upload.
```
I think you must have been looking at an older commit; all types in new code should be replaced by explicit annotations (which obviates the need for types in the docstrings)

austinbyers commented on lambda_functions/downloader/main.py:
```diff
@@ -7,115 +7,103 @@
-
-def _upload_to_s3(local_file_path, md5, observed_path):
-    """Upload a binary to S3, keyed by a UUID.
+    download_path = '/tmp/carbonblack_{}'.format(binary.md5)
```
Hmmm, that's a good point. Lambda explicitly allocates `/tmp` (which may not be the value returned by `gettempdir`). I'll try it and see!

austinbyers commented on lambda_functions/downloader/main.py:
```diff
@@ -0,0 +1,125 @@
+LOGGER = logging.getLogger()
+LOGGER.setLevel(logging.INFO)
+
+ENCRYPTED_TOKEN = os.environ['ENCRYPTED_CARBON_BLACK_API_TOKEN']
```
That's a good idea. I'm going to leave it for now because I want to standardize this across all of the Lambda functions in a subsequent PR

austinbyers replied:
For now, I will definitely add this check in `copy_all.py`, since that's the only code which is designed to be invoked by a local user

austinbyers commented on manage.py:
```diff
@@ -74,33 +72,102 @@ def help(self):
-        except ManagerError as error:
-            # Print error type and message, not full stack trace.
-            sys.exit('{}: {}'.format(type(error).__name__, error))
+        self._parse_config(allow_empty=(command == 'test'))
```
This was a confusing flag to help with validation; it has since been removed and it is hopefully clearer now

austinbyers commented on tests/lambda_functions/downloader/copy_all_test.py:
```diff
@@ -30,7 +30,7 @@ class MockMain(object):
     CARBON_BLACK = MockCarbonBlack()
 
-    def __init__(self, inject_errors: bool=False):
+    def __init__(self, inject_errors: bool = False):
```
Pylint actually complains if there is not the extra space:
```
************* Module tests.lambda_functions.downloader.copy_all_test
C: 39, 0: Exactly one space required around keyword argument assignment
    def __init__(self, inject_errors: bool=False):
                                          ^ (bad-whitespace)
```

However, this is not `keyword argument assignment`, so I'm guessing pylint doesn't exactly know how to handle type annotations?

I prefer the no-space version, but I don't want to have to disable the whitespace rule, so I guess I'll live with it :)

austinbyers replied:
Another note:
```python
def __init__(self, inject_errors=False: bool):
```
is invalid syntax

ryandeivert replied:
Thanks for clarifying!

austinbyers commented:
@ryandeivert PTAL

I've squashed the commits to simplify things and addressed all of the feedback. Be sure to look at the downloader code again because your last review was an an older commit for some reason

ryandeivert commented:
@austinbyers your'e right - I had been stepping through commits since this PR was so large :)

austinbyers commented:
@ryandeivert Thanks for your review! I should have done a better job of keeping track of and squashing the commits. I did another deploy just to test everything one last time and once the Travis tests pass I'll go ahead and merge
